### ISSUE TITLE
blue lines drawn with aalines washed out

### ISSUE THREAD

**Issue by benjaminbradley** [pygame/pygame#395]
ARGB Surfaces didn't work nicely with `draw.aaline`.

What's left to do:
- [x] a unit test or two in [test/draw_test.py](https://github.com/pygame/pygame/blob/master/test/draw_test.py#L166). Check ARGB and RGBA surfaces, with and without SRC_ALPHA. (**good first issue**)
- [x] check other platforms (the unit tests can do that for us mostly).
- [x] fix draw.c DRAWPIX32 to work with not RGBA format Surfaces.

---

Lines drawn with the pygame.draw.aalines() function seem to be washed out of the blue color. Here is a simple demo program to compare the color of lines drawn with pygame.draw.line() and pygame.draw.aalines(), both with the same color code, and a screencap of the difference on my monitor.

```
import pygame, colors, sys
image = pygame.display.set_mode((200,200))
image.fill((0,0,0))
pygame.draw.aalines(image, colors.BLUE, False, [(0,0), (100,100), (100, 200)])
pygame.draw.line(image, colors.BLUE, (0,100), (100,200), 2)
pygame.display.update()
while True:
	for event in pygame.event.get():
		if event.type == pygame.KEYDOWN and event.key == pygame.K_ESCAPE: sys.exit()
	pygame.time.Clock().tick(30)
```

![screen shot 2018-02-13 at 6 02 07 pm](https://user-images.githubusercontent.com/38849/36181099-e1abfe0e-10e8-11e8-9756-2940ce288660.png)

I'm running Python 3.6.2 and pygame 1.9.3 on Mac OS 10.12.6

**Comment by illume** [comment]
Confirmed on 10.11.x. The blue component is getting drawn as black.

```
import pygame, sys
screen = pygame.display.set_mode((200,200))
offscreen = screen.convert()
going = True

while going:
    for event in pygame.event.get():
        if ((event.type == pygame.KEYDOWN and event.key == pygame.K_ESCAPE) or
            event.type == pygame.QUIT):
            going = False

    screen.fill((0,0,0))
    offscreen.fill((0,0,0))
    if 0:
        pygame.draw.aalines(screen, (0, 0, 255), False, [(0,0), (100,100), (100, 200)])
        pygame.draw.line(screen, (0, 0, 255), (0,100), (100,200), 2)
    else:
        pygame.draw.aalines(offscreen, (255, 255, 255), False, [(0,0), (100,100), (100, 200)])
        pygame.draw.aalines(offscreen, (0, 255, 0), False, [(20,20), (0,50), (0, 50)])
        pygame.draw.aalines(offscreen, (0, 0, 255), False, [(30,30), (70,90), (80, 100)])
        pygame.draw.aaline(offscreen, (0, 0, 255), (40,40), (80,100))
        pygame.draw.line(offscreen, (0, 0, 255), (0,100), (100,200), 2)
        screen.blit(offscreen, (0,0))

    pygame.display.flip()
    pygame.time.Clock().tick(30)
```

**Comment by illume** [comment]
The next steps would be:
- [ ] a unit test or two in [test/draw_test.py](https://github.com/pygame/pygame/blob/master/test/draw_test.py#L166) 
- [ ] check other platforms (the unit tests can do that for us mostly).

**Comment by Connor124** [comment]
Slightly different on Debian 9 with Python 3.6.2 and pygame 1.9.3 with code above 

![uimage](https://user-images.githubusercontent.com/20603841/37306723-8335273a-2630-11e8-9e77-d16eac6c5dfc.png)

**Comment by illume** [comment]
- Seems an issue with surfaces that don't have alpha.
- bug also present in pygame 1.9.2.
- a work around is to draw onto a surface with alpha (`surf.convert_alpha()`)

Here's my above example, but also with `.convert_alpha()`.
```python
import pygame, sys
screen = pygame.display.set_mode((200,200))
offscreen = screen.convert_alpha()
going = True

while going:
    for event in pygame.event.get():
        if ((event.type == pygame.KEYDOWN and event.key == pygame.K_ESCAPE) or
            event.type == pygame.QUIT):
            going = False

    screen.fill((0,0,0))
    offscreen.fill((0,0,0))
    if 0:
        pygame.draw.aalines(screen, (0, 0, 255), False, [(0,0), (100,100), (100, 200)])
        pygame.draw.line(screen, (0, 0, 255), (0,100), (100,200), 2)
    else:
        pygame.draw.aalines(offscreen, (255, 255, 255), False, [(0,0), (100,100), (100, 200)])
        pygame.draw.aalines(offscreen, (0, 255, 0), False, [(20,20), (0,50), (0, 50)])
        pygame.draw.aalines(offscreen, (0, 0, 255), False, [(30,30), (70,90), (80, 100)])
        pygame.draw.aaline(offscreen, (0, 0, 255), (40,40), (80,100))
        pygame.draw.line(offscreen, (0, 0, 255), (0,100), (100,200), 2)
        screen.blit(offscreen, (0,0))

    pygame.display.flip()
    pygame.time.Clock().tick(30)


```

**Comment by illume** [comment]
There's a [DRAWPIX32 in draw.c](https://github.com/pygame/pygame/blob/master/src/draw.c#L936) which doesn't understand pixels and surface formats. It would need to be rewritten to support surfaces without transparency properly.

Rather than fail weirdly, I've made it raise an exception. So anyone who uses it knows it's limitations.

It would need to do something like this untested...

```c
int tmp_r, tmp_g, tmp_b, tmp_a;
float nbr;

SDL_GetRGBA(pixel, surf->format, &pixel_r, &pixel_g, &pixel_b, &pixel_a);
tmp_r = color_r * br + pixel_r * nbr;
tmp_g = color_g * br + pixel_g * nbr;
tmp_b = color_b * br + pixel_b * nbr;
*pixel = SDL_MapRGBA (surf->format, 
    (tmp_r>254) ? 255: tmp_r,
    (tmp_g>254) ? 255: tmp_g,
    (tmp_b>254) ? 255: tmp_b,
    (tmp_a>254) ? 255: tmp_a);
```
That's approximately the if(blend) branch of DRAWPIX32. 
https://wiki.libsdl.org/SDL_GetRGBA
https://wiki.libsdl.org/SDL_MapRGBA

**Comment by illume** [comment]
I updated the C function in #418

Now we just need a unit test to see if it actually works...

**Comment by illume** [comment]
This was fixed in 1.9.4, and there is a unit test now too. So, time to close this one.
